### ISSUE TITLE
Default exception handler not being called (PHP7)

### ISSUE THREAD

**Issue by BabakMN** [getsentry/sentry-php#407]
Not sure if it's expected or not. Sentry is swallowing all exceptions that it is catching without propagating them up to the native handler. As a result you get complete silence and nothing in the local logs.

`PHP 7.0.13-0ubuntu0.16.04.1`

```php
$error_handler = new \Raven_ErrorHandler($client);
$error_handler->registerExceptionHandler(true);
$error_handler->registerErrorHandler(true);
$error_handler->registerShutdownFunction();
```

Adding something like this before installing the handler makes it work.

```php
function root_exception_handler($exception) {
   throw $exception;
}

set_exception_handler('root_exception_handler');
```

Seems similar to #260.

**Comment by dcramer** [comment]
@BabakMN IIRC set_exception_handler functions differently than the others -- if you bind one, default behavior is no longer invoked. We could possibly restore it by looking for an existing handler (if we can, I forget), and doing what you're doing above.

**Comment by BabakMN** [comment]
I think most people would expect the default PHP handler to be called when the "call existing handler" parameter is true.

Right now it seems like it is actually behaving as "call any existing **custom** exception handlers".

So in terms of expected behaviour,

* if call_existing is false: only Sentry handles the exception, nothing else.

* if call_existing is true: Sentry and **also** the existing handler handle the exception, if a custom exception handler is not set, the native PHP exception handler is considered to be the existing handler and will be invoked.

When a custom exception handler is not installed `set_exception_handler` returns null so as a result we end up with silence (because `$this->old_exception_handler` becomes null).

How about we check for this case, and so,

* if (call_existing_handler is true) and (old_exception_handler is null): Let Sentry capture the exception, then re-throw it to invoke PHP's native handler as well.

How does that sound?

**Comment by dcramer** [comment]
I'm +1 on this approach. Haven't had time to do a PR myself yet but eventually will get around to it.

**Comment by BabakMN** [comment]
I'd be interested to work on a PR for this. Should I send one?

**Comment by dcramer** [comment]
@BabakMN If you have the time that'd be super helpful

**Comment by BabakMN** [comment]
How's this? https://github.com/BabakMN/sentry-php/commit/65021694c3e669aa3d1201281290223de0d6677e

It just re-throws if a custom exception handler is not set.

I noticed the error handler is behaving correctly by checking for null and returning false to propagate the error.

For tests it's a bit weird because we would have to register a custom exception handler to perform the test but then we can't test the case where no custom exception handler is set.

If we were to test it we'd have to find a way to inject a null into `old_exception_handler` to simulate that case and test it.

**Comment by dcramer** [comment]
This won't bubble it back up right? I think we have some protection for that just curious. Otherwise it looks sane to me. Also not sure how we'd automate tests here unless you explicitly called the handle function and confirmed it raised the exception passed. 






On Fri, Feb 24, 2017 at 5:29 AM -0800, "BabakMN" <notifications@github.com> wrote:












How's this? BabakMN@6502169


It just re-throws if a custom exception handler is not set.


I noticed the error handler is behaving correctly by checking for null and returning false to propagate the error.


For tests it's a bit weird because we would have to register a custom exception handler to perform the test but then we can't test the case where no custom exception handler is set.


If we were to test it we'd have to find a way to inject a null into old_exception_handler to simulate that case and test it.

**Comment by BabakMN** [comment]
You are right it wont bubble back up. It will cause a PHP fatal error (`Uncaught Exception`) and terminate the script which is expected behaviour (exception in exception handler).

It is expected because the user didn't have a custom exception handler and asked for them to be propagated (to the native handler).

And handing an exception to the native handler is the same as throwing it and not catching it, which causes the fatal error.

So in other words in this case Sentry becomes a passive observer of the exceptions but doesn't modify behaviour (such as swallowing exceptions).

I wrote a test based on your suggestion of explicitly calling the `handleException` function (https://github.com/BabakMN/sentry-php/commit/8c8f61c2fa01e7136630a1853fb060e1bc4f6d1a).

**Comment by dcramer** [comment]
Looks pretty good to me. Want to throw it up as a PR?

**Comment by BabakMN** [comment]
Thanks for checking. Yes I will send a PR soon.

**Comment by stayallive** [comment]
Looks like this was fixed by you in #421 so closing this issue 🎉
